File: src/main/scala/com/github/vickumar1981/stringdistance/StringDistance.scala
package com.github.vickumar1981.stringdistance

import com.github.vickumar1981.stringdistance.impl._

/**
  * Main class to organize functionality of different string distance algorithms
  *
  * {{{
  * import com.github.vickumar1981.stringdistance.Strategy
  * import com.github.vickumar1981.stringdistance.StringDistance._
  * import com.github.vickumar1981.stringdistance.impl.{ConstantGap, LinearGap}
  *
  * // Scores between strings
  * val cosSimilarity: Double = Cosine.score("hello", "chello", Strategy.splitWord)
  * val damerau: Double = Damerau.score("martha", "marhta")
  * val diceCoefficient: Double = DiceCoefficient.score("martha", "marhta")
  * val hamming: Double = Hamming.score("martha", "marhta")
  * val jaccard: Double = Jaccard.score("karolin", "kathrin", 1)
  * val jaro: Double = Jaro.score("martha", "marhta")
  * val jaroWinkler: Double = JaroWinkler.score("martha", "marhta", 0.1)
  * val levenshtein: Double = Levenshtein.score("martha", "marhta")
  * val needlemanWunsch: Double = NeedlemanWunsch.score("martha", "marhta", ConstantGap())
  * val ngramSimilarity: Double = NGram.score("karolin", "kathrin", 1)
  * val bigramSimilarity: Double = NGram.score("karolin", "kathrin", 2)
  * val overlap: Double = Overlap.score("karolin", "kathrin", 1)
  * val smithWaterman: Double = SmithWaterman.score("martha", "marhta", (LinearGap(gapValue = -1), Integer.MAX_VALUE))
  * val smithWatermanGotoh: Double = SmithWatermanGotoh.score("martha", "marhta", ConstantGap())
  * val tversky: Double = Tversky.score("karolin", "kathrin", 0.5)
  *
  * // Distances between strings
  * val damerauDist: Int = Damerau.distance("martha", "marhta")
  * val hammingDist: Int = Hamming.distance("martha", "marhta")
  * val levenshteinDist: Int = Levenshtein.distance("martha", "marhta")
  * val longestCommonSubSeq: Int = LongestCommonSeq.distance("martha", "marhta")
  * val ngramDist: Int = NGram.distance("karolin", "kathrin", 1)
  * val bigramDist: Int = NGram.distance("karolin", "kathrin", 2)
  * }}}
  */
object StringDistance {
  object Cosine extends WeightedStringMetric[CosineAlgorithm, String]
  object Damerau extends StringMetric[DamerauLevenshteinAlgorithm]
  object DiceCoefficient extends StringMetric[DiceCoefficientAlgorithm]
  object Hamming extends StringMetric[HammingAlgorithm]
  object Jaccard extends WeightedStringMetric[JaccardAlgorithm, Int]
  object Jaro extends StringMetric[JaroAlgorithm]
  object JaroWinkler extends WeightedStringMetric[JaroWinklerAlgorithm, Double]
  object Levenshtein extends StringMetric[LevenshteinAlgorithm]
  object LongestCommonSeq extends StringMetric[LongestCommonSeqAlorithm]
  object NeedlemanWunsch extends WeightedStringMetric[NeedlemanWunschAlgorithm, ConstantGap]
  object NGram extends WeightedStringMetric[NGramAlgorithm, Int]
  object Overlap extends WeightedStringMetric[OverlapAlgorithm, Int]
  object SmithWaterman extends WeightedStringMetric[SmithWatermanAlgorithm, (Gap, Int)]
  object SmithWatermanGotoh extends WeightedStringMetric[SmithWatermanGotohAlgorithm, ConstantGap]
  object Tversky extends WeightedStringMetric[TverskyAlgorithm, Double]
}

/**
  * Jave Wrapper for cosine similarity.
  */
class CosineSimilarityImplWrapper extends CosSimilarityImpl

/**
  * Jave Wrapper for dice coefficient similarity.
  */
class DiceCoefficientImplWrapper extends DiceCoefficientImpl

/**
  * Jave Wrapper for hamming distance.
  */
class HammingImplWrapper extends HammingImpl

/**
  * Jave Wrapper for jaccard similarity.
  */
class JaccardImplWrapper extends JaccardImpl

/**
  * Jave Wrapper for jaro and jaro winkler similarity.
  */
class JaroImplWrapper extends JaroImpl

/**
  * Jave Wrapper for levenshtein distance.
  */
class LevenshteinDistanceImplWrapper extends LevenshteinDistanceImpl[Char]

/**
  * Jave Wrapper for longest comment sequence.
  */
class LongestCommonSeqWrapper extends LongestCommonSeqImpl

/**
  * Jave Wrapper for needleman wunsch similarity.
  */
class NeedlemanWunschImplWrapper extends NeedlemanWunschImpl

/**
  * Jave Wrapper for n-gram similarity.
  */
class NGramImplWrapper extends NGramImpl

/**
  * Jave Wrapper for overlap similarity.
  */
class OverlapImplWrapper extends OverlapImpl


/**
  * Jave Wrapper for smith waterman similarity.
  */
class SmithWatermanImplWrapper extends SmithWatermanImpl

File: src/main/java/com/github/vickumar1981/stringdistance/util/StringDistance.java
package com.github.vickumar1981.stringdistance.util;

import com.github.vickumar1981.stringdistance.impl.ConstantGap;
import com.github.vickumar1981.stringdistance.impl.Gap;
import com.github.vickumar1981.stringdistance.impl.LinearGap;

public class StringDistance {
    private final static CosSimilarityImpl cosine = new CosSimilarityImpl();
    private final static JaccardImpl jaccard = new JaccardImpl();
    private final static JaroImpl jaro = new JaroImpl();
    private final static HammingImpl hamming = new HammingImpl();
    private final static DiceCoefficientImpl dice = new DiceCoefficientImpl();
    private final static LevenshteinDistanceImpl levenshtein = new LevenshteinDistanceImpl();
    private final static LongestCommonSeqImpl longestCommonSeq = new LongestCommonSeqImpl();
    private final static NeedlemanWunschImpl needlemanWunsch = new NeedlemanWunschImpl();
    private final static NGramImpl ngram = new NGramImpl();
    private final static OverlapImpl overLap = new OverlapImpl();
    private final static SmithWatermanImpl smithWaterman = new SmithWatermanImpl();

    private final static String splitOnWord = "(?!^)";
    private final static String splitOnSentence = "\\W+";

    private static Double scoreFromDistance(String s1, String s2, Integer distance) {
        int maxLen = Math.max(s1.length(), s2.length());
        int minLen = maxLen - distance;
        if (minLen < 0 || minLen > maxLen) {
            minLen = 0;
        }
        return minLen * 1.0 / maxLen;
    }

    public static Double cosine(String s1, String s2) {
        return cosine(s1, s2, splitOnWord);
    }

    public static Double cosine(String s1, String s2, String splitOn) {
        return cosine.cosSimilarity(s1, s2, splitOn);
    }

    public static Double damerau(String s1, String s2) {
        return scoreFromDistance(s1, s2, damerauDist(s1, s2));
    }

    public static Integer damerauDist(String s1, String s2) {
        return levenshtein.damerauLevenshtein(s1, s2);
    }

    public static Double diceCoefficient(String s1, String s2) {
        return dice.diceCoefficient(s1, s2);
    }

    public static Double hamming(String s1, String s2) {
        return scoreFromDistance(s1, s2, hammingDist(s1, s2));
    }

    public static Integer hammingDist(String s1, String s2) {
        return hamming.hamming(s1, s2);
    }

    public static Double jaccard(String s1, String s2) { return jaccard(s1, s2, 1); }

    public static Double jaccard(String s1, String s2, Integer n) { return jaccard.jaccard(s1, s2, n); }

    public static Double jaroWinkler(String s1, String s2) {
        return jaroWinkler(s1, s2, 0.1);
    }

    public static Double jaroWinkler(String s1, String s2, Double weight) {
        return jaro.jaroWinkler(s1, s2, weight);
    }

    public static Double jaro(String s1, String s2) {
        return jaro.jaro(s1, s2);
    }

    public static Double levenshtein(String s1, String s2) {
        return scoreFromDistance(s1, s2, levenshteinDist(s1, s2));
    }

    public static Integer levenshteinDist(String s1, String s2) {
        return levenshtein.levenshtein(s1, s2);
    }

    public static Integer longestCommonSeq(String s1, String s2) {
        return longestCommonSeq.longestCommonSeq(s1, s2);
    }

    public static Double needlemanWunsch(String s1, String s2) {
        return needlemanWunsch(s1, s2, new ConstantGap(1d, -1d, 0d));
    }

    public static Double needlemanWunsch(String s1, String s2, ConstantGap gap) {
            return needlemanWunsch.needleman(s1, s2, gap);
    }

    public static Integer nGramDist(String s1, String s2) { return nGramDist(s1, s2, 1); }

    public static Integer nGramDist(String s1, String s2, Integer n) { return ngram.nGramDist(s1, s2, n); }

    public static Double nGram(String s1, String s2) { return nGram(s1, s2, 1); }

    public static Double nGram(String s1, String s2, Integer n) { return ngram.nGram(s1, s2, n); }

    public static Double overlap(String s1, String s2) { return overlap(s1, s2, 1); }

    public static Double overlap(String s1, String s2, Integer n) { return overLap.overlap(s1, s2, n); }

    public static Double smithWaterman(String s1, String s2) {
        return smithWaterman(s1, s2, new LinearGap(1d, -1d, -1d));
    }

    public static Double smithWaterman(String s1, String s2, Gap gap) {
        return smithWaterman.smithWaterman(s1, s2, gap, Integer.MAX_VALUE);
    }

    public static Double smithWaterman(String s1, String s2, Gap gap, Integer windowSize) {
        return smithWaterman.smithWaterman(s1, s2, gap, windowSize);
    }

    public static Double smithWatermanGotoh(String s1, String s2) {
        return smithWatermanGotoh(s1, s2, new ConstantGap(1d, -1d, -1d));
    }

    public static Double smithWatermanGotoh(String s1, String s2, ConstantGap gap) {
        return smithWaterman.smithWatermanGotoh(s1, s2, gap);
    }

    public static Double tversky(String s1, String s2) { return tversky(s1, s2, 1d); }

    public static Double tversky(String s1, String s2, Double n) { return jaccard.tversky(s1, s2, n); }
}
